resources/tests/org.eclipse.core.tests.resources/src/org/eclipse/core/tests/internal/builders/AutoBuildJobRescheduleTest.java

```diff
@@ -0,0 +1,200 @@
+		try {
+			// Turn auto-building off to prevent too early builds
+			setAutoBuilding(false);
+			// Create and open the project
+			project.create(getMonitor());
+			project.open(getMonitor());
+			// Create the file
+			ensureExistsInWorkspace(file, getRandomContents());
+
+			// Add the EmptyDeltaBuilder. The behavior of that builder is configured below.
+			IProjectDescription desc = project.getDescription();
+			desc.setBuildSpec(new ICommand[] { createCommand(desc, EmptyDeltaBuilder.BUILDER_NAME, "ProjectBuild") });
+			project.setDescription(desc, getMonitor());
+			project.build(IncrementalProjectBuilder.INCREMENTAL_BUILD, getMonitor());
+
+			// Perform the first build. This instantiates the EmptyDeltaBuilder and makes
+			// its instance accessible.
+			setAutoBuilding(true);
+
+			// Configure the behavior of the builder
+			EmptyDeltaBuilder.getInstance().setRuleCallback(new BuilderRuleCallback() {
+				// We only execute the block-continue on the first run in the test case.
+				// Therefore, we remember whether we have already run with this flag.
+				private AtomicBoolean buildRan = new AtomicBoolean(false);
+
+				@Override
+				public IProject[] build(int kind, Map<String, String> args, IProgressMonitor monitor)
+						throws CoreException {
+					// only on the first run, block the builder
+					if (!buildRan.getAndSet(true)) {
+						barrier.setStatus(AUTOBUILD_RUNNING_BLOCKED);
+						// block until the WorkspaceJob has made the file dirty
+						barrier.waitForStatus(AUTOBUILD_CONTINUE);
+					}
+					return new IProject[] { project };
+				}
+			});
+
+			// Wait until the Content Description Update job has finished, because it also
+			// might schedule a build and we don't want that interfere with our test
+			// process.
+			waitForContentDescriptionUpdate();
+		} catch (CoreException e) {
+			fail("Unexpected exception in project setup", e);
+		}
+	}
+
+	@Override
+	protected void tearDown() throws Exception {
+		Job.getJobManager().removeJobChangeListener(jobChangeListener);
+
+		super.tearDown();
+	}
+
+	public void testAutoBuildRescheduling() throws InterruptedException, CoreException {
+		// trigger an autobuild. This will execute the builder we have configured above
+		// to block.
+		dirty(file);
+
+		// Let a concurrent job modify the file
+		var job = new WorkspaceJob("ModifyFile") {
+			@Override
+			public IStatus runInWorkspace(IProgressMonitor monitor) throws CoreException {
+				// Make sure the builder is blocking
+				barrier.waitForStatus(AUTOBUILD_RUNNING_BLOCKED);
+				// Now modify the file - this requests another build
+				dirty(file);
+				// Register our jobChangeListener so we can record when the AutoBuildJob is
+				// rescheduled and done.
+				Job.getJobManager().addJobChangeListener(jobChangeListener);
+				// Signal to the builder that it can continue now
+				barrier.setStatus(AUTOBUILD_CONTINUE);
+				return Status.OK_STATUS;
+			}
+		};
+		job.setRule(new SpoiledSchedulingRule());
```

Comment:
I do not understand that rule. what should it do other then just no rule? and why do you schedule and join the job instead of having no job at all and just runinworkspace?

Reply:
For the job/runnable itself, there is no difference. But dirty(file) internally will create a rule and will block on the AutoBuildJob. The `SpoiledSchedulingRule` behaves like a null rule for the job/runnable but still contains the rule for `file` which causes the dirty operation to not block on the AutoBuildJob.

Reply:
And about the `Workspace.run()` you are right, I was thinking too complex here...